whelktool/src/main/groovy/whelk/datatool/form/Transform.groovy

```diff
@@ -182,6 +302,133 @@ class Transform {
+    }
+
+    private void putIds(Map form) {
+        DocumentUtil.traverse(form) { value, path ->
+            if (asList(value).isEmpty()) {
+                return new DocumentUtil.Replace([(ID_KEY): EMPTY_BLANK_NODE_TMP_ID])
+            } else if (value instanceof Map) {
+                def _id = value.remove(_ID)
+                if (nodeIdMappings.containsKey(_id)) {
+                    value[ID_KEY] = _id
+                    return new DocumentUtil.Nop()
+                } else if (value.isEmpty()) {
+                    return new DocumentUtil.Replace([(ID_KEY): EMPTY_BLANK_NODE_TMP_ID])
+                }
+            }
+        }
+    }
+
+    private static String withoutPrefixes(String ttl) {
+        ttl.readLines()
+                .split { it.startsWith('prefix') }
+                .get(1)
+                .join('\n')
+                .trim()
+    }
+
+    Map<String, List<String>> collectNodeIdMappings(Whelk whelk) {
+        Map<String, List<String>> nodeIdMappings = [:]
+
+        IdLoader idLoader = whelk ? new IdLoader(whelk.storage) : null
+
+        DocumentUtil.traverse(matchForm) { value, path ->
+            if (value instanceof Map && path && dropLastIndex(path).last() == _ID_LIST) {
+                def ids = value[VALUE] as List<String>
+                        ?: (value[VALUE_FROM] ? IdLoader.fromFile((String) value[VALUE_FROM][ID_KEY]) : [])
+                if (ids) {
+                    def node = getAtPath(matchForm, dropLastIndex(path).dropRight(1))
+                    def nodeId = (String) node[_ID]
+
+                    if (!idLoader) {
+                        nodeIdMappings[nodeId] = ids
+                        return
+                    }
+
+                    def (iris, shortIds) = ids.split(JsonLd::looksLikeIri)
+                    // Assume given http strings to be valid iris and thus avoid the round trip of first finding the
+                    // short id (in lddb__identifiers) and then use the short id only to find (most likely) the same
+                    // thing/record iri (in lddb) that we started with.
+                    if (shortIds.isEmpty()) {
+                        nodeIdMappings[nodeId] = iris
+                        return
+                    }
+
+                    def nodeType = node[TYPE_KEY]
+                    def marcCollection = nodeType ? getMarcCollectionInHierarchy((String) nodeType, whelk.jsonld) : null
+                    def xlShortIds = idLoader.collectXlShortIds(shortIds as List<String>, marcCollection)
+                    def parentProp = dropIndexes(path).reverse()[1]
+                    def isInRange = { type -> whelk.jsonld.getInRange(type).contains(parentProp) }
+                    // TODO: Fix hardcoding
+                    def isRecord = whelk.jsonld.isInstanceOf((Map) node, "AdminMetadata")
+                            || isInRange(RECORD_TYPE)
+                            || isInRange("AdminMetadata")
```

Comment:
AdminMetadata is a platform term so I think it is fine.
But it could be defined in `JsonLd`